chain/store/store.go

```diff
@@ -354,27 +376,32 @@ func (cs *ChainStore) persist(b *block.Block) error {
 	}
 
 	//StateRoot
-	states, root, err := cs.generateStateRoot(context.Background(), b, b.Header.UnsignedHeader.Height != 0, true)
-	if err != nil {
-		return err
-	}
+	if !cs.DuringFastSync(b.Header.UnsignedHeader.Height) {
+		states, root, err := cs.generateStateRoot(context.Background(), b, b.Header.UnsignedHeader.Height != 0, true)
+		if err != nil {
+			return err
+		}
 
-	headerRoot, err := common.Uint256ParseFromBytes(b.Header.UnsignedHeader.StateRoot)
-	if err != nil {
-		return err
-	}
-	if ok := root.CompareTo(headerRoot); ok != 0 {
-		return fmt.Errorf("state root not equal:%v, %v", root.ToHexString(), headerRoot.ToHexString())
-	}
+		headerRoot, err := common.Uint256ParseFromBytes(b.Header.UnsignedHeader.StateRoot)
+		if err != nil {
+			return err
+		}
+		if ok := root.CompareTo(headerRoot); ok != 0 {
+			return fmt.Errorf("state root not equal:%v, %v", root.ToHexString(), headerRoot.ToHexString())
+		}
 
-	err = cs.st.BatchPut(db.CurrentStateTrie(), root.ToArray())
-	if err != nil {
-		return err
+		err = cs.st.BatchPut(db.CurrentStateTrie(), root.ToArray())
+		if err != nil {
+			return err
+		}
+		defer func() {
+			cs.States = states
```

Comment:
Why move it to here? If db write failed (`Commit()` return error), `cs.states` should not be updated.